### ISSUE TITLE
Suggestion of lint `clippy::map_entry` causes error

### ISSUE THREAD

**Issue by viadezo1er** [rust-lang/rust-clippy#17081]
### Summary

https://github.com/braintrustdata/bt
commit 14b3ec21faadf730bd199173811a1a28798babcb

`RUST_BACKTRACE=1 cargo clippy --fix --bin "bt" -p bt -- --no-deps`

### Version

```text
rustc 1.95.0 (59807616e 2026-04-14) (Homebrew)
binary: rustc
commit-hash: 59807616e1fa2540724bfbac14d7976d7e4a3860
commit-date: 2026-04-14
host: aarch64-apple-darwin
release: 1.95.0
LLVM version: 22.1.3
```

### Error output

<details><summary>Backtrace</summary>
  <p>

  ```
      Checking bt v0.10.0 (/Users/cedric@braintrustdata.com/repos/bt-fix-ci)
warning: failed to automatically apply fixes suggested by rustc to crate `bt`

after fixes were automatically applied the compiler reported errors within these files:

  * src/sync/discovery.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0382]: borrow of moved value: `root_span_id`
   --> src/sync/discovery.rs:101:42
    |
 96 |                     let root_span_id = reference.root_span_id.clone();
    |                         ------------ move occurs because `root_span_id` has type `std::string::String`, which does not implement the `Copy` trait
 97 |                     if let std::collections::hash_map::Entry::Vacant(e) = trace_refs_by_root_span_id.entry(root_span_id) {
    |                                                                                                            ------------ value moved here
...
101 |                         trace_roots.push(root_span_id.clone());
    |                                          ^^^^^^^^^^^^ value borrowed here after move

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0382`.
Original diagnostics will follow.

warning: this function has too many arguments (8/7)
    --> src/datasets/pipeline.rs:1259:1
     |
1259 | / async fn transform_source_refs(
1260 | |     base: &BaseArgs,
1261 | |     runner: &PipelineRunnerArgs,
1262 | |     source_project_id: &str,
...    |
1267 | |     mut row_writer: Option<&mut dyn Write>,
1268 | | ) -> Result<PipelineTransformResponse> {
     | |______________________________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#too_many_arguments
     = note: `#[warn(clippy::too_many_arguments)]` on by default

warning: this `if` can be collapsed into the outer `match`
    --> src/datasets/pipeline.rs:2009:17
     |
2009 | /                 if *byte != b'-' {
2010 | |                     return false;
2011 | |                 }
     | |_________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
     = note: `#[warn(clippy::collapsible_match)]` on by default
help: collapse nested if block
     |
2008 ~             8 | 13 | 18 | 23
2009 ~                 if *byte != b'-' => {
2010 |                     return false;
2011 ~                 }
     |

warning: consider using `sort_by_key`
    --> src/functions/pull.rs:1413:13
     |
1413 |             entries.sort_by(|(left, _), (right, _)| left.cmp(right));
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#unnecessary_sort_by
     = note: `#[warn(clippy::unnecessary_sort_by)]` on by default
help: try
     |
1413 -             entries.sort_by(|(left, _), (right, _)| left.cmp(right));
1413 +             entries.sort_by_key(|(left, _)| *left);
     |

warning: this `if` can be collapsed into the outer `match`
   --> src/sql.rs:202:21
    |
202 | /                     if handle_key_event(&mut app, key, &client, &handle)? {
203 | |                         break;
204 | |                     }
    | |_____________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
    |
201 ~                 Event::Key(key)
202 ~                     if handle_key_event(&mut app, key, &client, &handle)? => {
203 |                         break;
204 ~                     }
    |

warning: this `if` can be collapsed into the outer `match`
   --> src/sql.rs:262:13
    |
262 | /             if !key.modifiers.contains(KeyModifiers::ALT) {
263 | |                 app.insert_char(ch);
264 | |             }
    | |_____________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
    |
261 ~         KeyCode::Char(ch) if !key.modifiers.contains(KeyModifiers::CONTROL)
262 ~             && !key.modifiers.contains(KeyModifiers::ALT) => {
263 |                 app.insert_char(ch);
264 ~             }
    |

warning: this function has too many arguments (8/7)
  --> src/sync/discovery.rs:56:1
   |
56 | / pub(crate) async fn discover_project_log_refs<F>(
57 | |     client: &ApiClient,
58 | |     ctx: &LoginContext,
59 | |     project_id: &str,
...  |
66 | | where
67 | |     F: FnMut(ProjectLogRef) -> Result<()>,
   | |__________________________________________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#too_many_arguments

warning: usage of `contains_key` followed by `insert` on a `HashMap`
   --> src/sync/discovery.rs:97:21
    |
 97 | /                     if !trace_refs_by_root_span_id.contains_key(&root_span_id) {
 98 | |                         if trace_roots.len() >= target {
 99 | |                             continue;
...   |
103 | |                         continue;
104 | |                     }
    | |_____________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#map_entry
    = note: `#[warn(clippy::map_entry)]` on by default
help: try
    |
 97 ~                     if let std::collections::hash_map::Entry::Vacant(e) = trace_refs_by_root_span_id.entry(root_span_id) {
 98 +                         if trace_roots.len() >= target {
 99 +                             continue;
100 +                         }
101 +                         trace_roots.push(root_span_id.clone());
102 +                         e.insert(reference);
103 +                         continue;
104 +                     }
    |

warning: large size difference between variants
   --> src/topics/mod.rs:194:1
    |
194 | / enum TopicMapCommands {
195 | |     /// Show a configured Topics topic map by name or function ID
196 | |     #[command(alias = "view")]
197 | |     Show(TopicMapViewArgs),
    | |     ---------------------- the second-largest variant contains at least 48 bytes
198 | |     /// Update a configured Topics topic map by name or function ID
199 | |     Set(TopicMapSetArgs),
    | |     -------------------- the largest variant contains at least 304 bytes
200 | | }
    | |_^ the entire enum is at least 304 bytes
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#large_enum_variant
    = note: `#[warn(clippy::large_enum_variant)]` on by default
help: consider boxing the large fields or introducing indirection in some other way to reduce the total size of the enum
    |
199 -     Set(TopicMapSetArgs),
199 +     Set(Box<TopicMapSetArgs>),
    |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:1543:21
     |
1543 | /                     if handle_key_event(app, key, &client, &handle)? {
1544 | |                         break;
1545 | |                     }
     | |_____________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
1542 ~                 Event::Key(key)
1543 ~                     if handle_key_event(app, key, &client, &handle)? => {
1544 |                         break;
1545 ~                     }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:1965:13
     |
1965 | /             if app.detail_pane_focus == DetailPaneFocus::Detail {
1966 | |                 if detail_message_list_mode {
1967 | |                     if app.detail_view == DetailView::Thread {
1968 | |                         for _ in 0..10 {
...    |
1979 | |             }
     | |_____________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
1964 ~         KeyCode::PageUp
1965 ~             if app.detail_pane_focus == DetailPaneFocus::Detail => {
1966 |                 if detail_message_list_mode {
 ...
1978 |                 }
1979 ~             }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:1982:13
     |
1982 | /             if app.detail_pane_focus == DetailPaneFocus::Detail {
1983 | |                 if detail_message_list_mode {
1984 | |                     if app.detail_view == DetailView::Thread {
1985 | |                         for _ in 0..10 {
...    |
1996 | |             }
     | |_____________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
1981 ~         KeyCode::PageDown
1982 ~             if app.detail_pane_focus == DetailPaneFocus::Detail => {
1983 |                 if detail_message_list_mode {
 ...
1995 |                 }
1996 ~             }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:2030:13
     |
2030 | /             if detail_message_list_mode {
2031 | |                 app.open_message_viewer();
2032 | |                 app.set_status(app.span_detail_status_text());
2033 | |             }
     | |_____________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
2029 ~         KeyCode::Char('v')
2030 ~             if detail_message_list_mode => {
2031 |                 app.open_message_viewer();
2032 |                 app.set_status(app.span_detail_status_text());
2033 ~             }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:2053:13
     |
2053 | /             if app.detail_view == DetailView::Span {
2054 | |                 app.toggle_span_detail_tab();
2055 | |                 app.set_status(app.span_detail_status_text());
2056 | |             }
     | |_____________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
2052 ~         KeyCode::Tab
2053 ~             if app.detail_view == DetailView::Span => {
2054 |                 app.toggle_span_detail_tab();
2055 |                 app.set_status(app.span_detail_status_text());
2056 ~             }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:2059:13
     |
2059 | /             if app.loaded_root_span_id.is_some() {
2060 | |                 request_loaded_trace_reload(app, client, handle)?;
2061 | |             }
     | |_____________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
2058 ~         KeyCode::Char('r')
2059 ~             if app.loaded_root_span_id.is_some() => {
2060 |                 request_loaded_trace_reload(app, client, handle)?;
2061 ~             }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:2064:13
     |
2064 | /             if app.detail_pane_focus == DetailPaneFocus::Tree
2065 | |                 && app.toggle_selected_span_collapsed()
2066 | |             {
2067 | |                 app.set_status(app.span_detail_status_text());
2068 | |             }
     | |_____________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
2063 ~         KeyCode::Char(' ')
2064 |             if app.detail_pane_focus == DetailPaneFocus::Tree
2065 |                 && app.toggle_selected_span_collapsed()
2066 ~             => {
2067 |                 app.set_status(app.span_detail_status_text());
2068 ~             }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:2071:13
     |
2071 | /             if app.detail_pane_focus == DetailPaneFocus::Detail {
2072 | |                 if detail_message_list_mode {
2073 | |                     if app.detail_view == DetailView::Thread {
2074 | |                         for _ in 0..10 {
...    |
2085 | |             }
     | |_____________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
2070 ~         KeyCode::Char('u') if key.modifiers.contains(KeyModifiers::CONTROL)
2071 ~             && app.detail_pane_focus == DetailPaneFocus::Detail => {
2072 |                 if detail_message_list_mode {
 ...
2084 |                 }
2085 ~             }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:2088:13
     |
2088 | /             if app.detail_pane_focus == DetailPaneFocus::Detail {
2089 | |                 if detail_message_list_mode {
2090 | |                     if app.detail_view == DetailView::Thread {
2091 | |                         for _ in 0..10 {
...    |
2102 | |             }
     | |_____________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
2087 ~         KeyCode::Char('d') if key.modifiers.contains(KeyModifiers::CONTROL)
2088 ~             && app.detail_pane_focus == DetailPaneFocus::Detail => {
2089 |                 if detail_message_list_mode {
 ...
2101 |                 }
2102 ~             }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:5223:17
     |
5223 | /                 if !value.is_empty() {
5224 | |                     parsed.row_ref = Some(value.to_string());
5225 | |                 }
     | |_________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
5222 ~             "r"
5223 ~                 if !value.is_empty() => {
5224 |                     parsed.row_ref = Some(value.to_string());
5225 ~                 }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:5228:17
     |
5228 | /                 if !value.is_empty() {
5229 | |                     parsed.span_id = Some(value.to_string());
5230 | |                 }
     | |_________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
5227 ~             "s"
5228 ~                 if !value.is_empty() => {
5229 |                     parsed.span_id = Some(value.to_string());
5230 ~                 }
     |

warning: this `if` can be collapsed into the outer `match`
    --> src/traces.rs:5233:17
     |
5233 | /                 if !value.is_empty() {
5234 | |                     parsed.trace_view_type = Some(value.to_string());
5235 | |                 }
     | |_________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
     |
5232 ~             "tvt"
5233 ~                 if !value.is_empty() => {
5234 |                     parsed.trace_view_type = Some(value.to_string());
5235 ~                 }
     |

warning: `bt` (bin "bt") generated 20 warnings (run `cargo clippy --fix --bin "bt" -p bt -- --no-deps` to apply 17 suggestions)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 3.59s

  ```

  </p>
</details>

**Comment by flip1995** [comment]
This is not an ICE, but some lint suggestion causes an error. ~~With a quick look, I couldn't identify which lint exactly. It would be really helpful if you could check which lint suggested the code change that caused this.~~

EDIT it's `clippy::map_entry` with this suggestion:


```
warning: usage of `contains_key` followed by `insert` on a `HashMap`
 --> src/sync/discovery.rs:97:21
  |
97 | /                     if !trace_refs_by_root_span_id.contains_key(&root_span_id) {
98 | |                         if trace_roots.len() >= target {
99 | |                             continue;
...   |
103 | |                         continue;
104 | |                     }
  | |_____________________^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#map_entry
  = note: `#[warn(clippy::map_entry)]` on by default
help: try
  |
97 ~                     if let std::collections::hash_map::Entry::Vacant(e) = trace_refs_by_root_span_id.entry(root_span_id) {
98 +                         if trace_roots.len() >= target {
99 +                             continue;
100 +                         }
101 +                         trace_roots.push(root_span_id.clone());
102 +                         e.insert(reference);
103 +                         continue;
104 +                     }
  |
```

**Comment by viadezo1er** [comment]
I reported this error since `cargo clippy` asked me to, I can help if there's something I can do but otherwise I'm not really planning to get involved in the fix.
To be clear this crash is not an issue for me.

Also sorry for the wrong label 😅

**Comment by flip1995** [comment]
If you feel like it, you can try to reduce this to a minimal reproducer. That usually helps most. That would involve extracting this snippet of the code in a way that it still triggers the lint and the suggestion still produces wrong code.

No worries about the label :)

**Comment by viadezo1er** [comment]
`Cargo.toml`:
```toml
[package]
name = "bt"
version = "0.0.0"
edition = "2021"

[[bin]]
name = "bt"
path = "src/main.rs"
```

`Cargo.lock`:
```toml
# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 4

[[package]]
name = "bt"
version = "0.0.0"
```

`src/main.rs`:
```rust
use std::collections::HashMap;

fn main() {
    let mut map: HashMap<String, i32> = HashMap::new();
    let mut keys: Vec<String> = Vec::new();
    let key = String::from("hello");

    if !map.contains_key(&key) {
        keys.push(key.clone());
        map.insert(key, 1);
    }

    println!("{:?} {:?}", map, keys);
}
```

The diff after running `RUST_BACKTRACE=1 cargo clippy --fix --broken-code --bin "bt" -p bt -- --no-deps`:
```diff
diff --git a/src/main.rs b/src/main.rs
index 5d0b95e..9d76d53 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -5,10 +5,10 @@ fn main() {
     let mut keys: Vec<String> = Vec::new();
     let key = String::from("hello");
 
-    if !map.contains_key(&key) {
+    map.entry(key).or_insert_with(|| {
         keys.push(key.clone());
-        map.insert(key, 1);
-    }
+        1
+    });
 
     println!("{:?} {:?}", map, keys);
 }
```

<details><summary>Backtrace</summary>
  <p>

```
    Checking bt v0.0.0 (/Users/cedric@braintrustdata.com/repos/bt-clippy-issue)
warning: failed to automatically apply fixes suggested by rustc to crate `bt`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0382]: borrow of moved value: `key`
 --> src/main.rs:8:35
  |
6 |     let key = String::from("hello");
  |         --- move occurs because `key` has type `std::string::String`, which does not implement the `Copy` trait
7 |
8 |     map.entry(key).or_insert_with(|| {
  |               ---                 ^^ value borrowed here after move
  |               |
  |               value moved here
9 |         keys.push(key.clone());
  |                   --- borrow occurs due to use in closure
  |
help: consider cloning the value if the performance cost is acceptable
  |
8 |     map.entry(key.clone()).or_insert_with(|| {
  |                  ++++++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0382`.
Original diagnostics will follow.

warning: usage of `contains_key` followed by `insert` on a `HashMap`
  --> src/main.rs:8:5
   |
 8 | /     if !map.contains_key(&key) {
 9 | |         keys.push(key.clone());
10 | |         map.insert(key, 1);
11 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#map_entry
   = note: `#[warn(clippy::map_entry)]` on by default
help: try
   |
 8 ~     map.entry(key).or_insert_with(|| {
 9 +         keys.push(key.clone());
10 +         1
11 +     });
   |

warning: `bt` (bin "bt") generated 1 warning (run `cargo clippy --fix --bin "bt" -p bt -- --no-deps` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.08s

```
  </p>
</details>

**Comment by flip1995** [comment]
I guess the correct suggestion would be

```
    map.entry(key.clone()).or_insert_with(|| {
       keys.push(key);
       1
    });
```

So, moving the `.clone()` from inside the body to the `key` in the `entry`.

I think if the key that is checked is itself not `Copy` and cloned inside of the `if` body, we should change the suggestion to suggest `entry(key.clone())`. And if the key is itself not cloned in the `map.insert` inside of the `if` body, we can additionally suggest to remove the last clone of the key in the body.

---

To summarize, I'd say the lint is correct in triggering here. The suggestion is just not good enough for this case to be auto-applicable.
